entc/gen/template/builder/mutation.tmpl

```diff
@@ -172,7 +172,10 @@ func (m {{ $mutation }}) Tx() (*Tx, error) {
 		switch {
 		case m.op.Is(OpUpdateOne | OpDeleteOne):
 			id, exists := m.ID()
-			if exists {
+			if exists && len(m.predicates) > 0 {
+				m.predicates = append(m.predicates, {{ $n.Package }}.ID(id))
+				return m.Client().{{ $n.Name }}.Query().Where(m.predicates...).IDs(ctx)
```

Comment:
This returns zero IDs in case the extra predicates are falsey. The extra predicates are respected/evaluated in a different layer of the codegen: https://github.com/ent/ent/blob/master/entc/gen/template/dialect/sql/update.tmpl#L86-L92

Reply:
That's exactly the intention behind this PR :) Take a look at the issue and example repo in the issue.
When using privacy filters, they are not applied when using the IDs() method inside hooks for UpdateOne and DeleteOne mutations.

This leads to a simple issue. E.g. a privacy filter is defined so that only users that match certain criteria can be mutated. Now there is a hook running OnDelete|OnDeleteOne that should do some cleanup and therefore simply takes the IDs() of the mutation to see which items should be deleted. Although the mutation will not run, the hook will be called with the id of the entity leading to stuff getting cleaned up that souldn't be. 

In our concrete use-case the cleanup is non-reversible due to it calling external APIs and is triggered by another hook on an edge related to the user which requires the user to still exist. So mutating first in the hook and checking for an error is not possible as well as running everything inside a transaction because of the external factor.

In my opinion one should expect the `IDs()` call to work the same way independent of being called from Delete vs. DeleteOne. This way hooks can be better reused and code duplication can be avoided.

Reply:
Additionally it makes the function behave as it is commented: 
"// IDs queries the database and returns the entity ids **that match the mutation's predicate**."

For a single ID it should be possible to not match the predicate of the mutation resulting in an empty list of IDs.

Reply:
Hey @mwoelk, I didn't miss the previous comment, and there's no need to "convince" me that the new logic is correct :). I just need to test some large codebases to ensure this change doesn't break any existing behavior. I will update you soon.

Reply:
Can you please add test cases for this? You can add them under the `integration/privacy` package.

Reply:
Sorry. Have been away a few days. Will add some tomorrow.

Reply:
Done. Added a new public "Note" entity that can be set to readonly during creation. A filter is then applied during mutation to filter out the ones set to readonly. Added some tests to `privacy_test.go` for UpdateOne, DeleteOne for both filtered (readonly) and unfiltered (not-readonly) notes and test for Update and Delete many. The tests inject a hook that compares present IDs to expected ones.

Reply:
@a8m let me know if I should add more test cases or if I can help getting this merged in any way